floris/wind_data.py

```diff
@@ -452,25 +452,51 @@ def upsample(self, wd_step=None, ws_step=None, method="linear", inplace=False):
             )
 
+        # First establish the current ws_step and wd_step
+        if len(self.wind_speeds) >= 2:
+            ws_step_current = self.wind_speeds[1] - self.wind_speeds[0]
+        else:  # wind rose will have only a single wind speed, and we assume a ws_step of 1
+            ws_step_current = 1.0
+
+        if len(self.wind_directions) >= 2:
+            wd_step_current = self.wind_directions[1] - self.wind_directions[0]
+        else:  # wind rose will have only a single wind direction, and we assume a wd_step of 1
+            wd_step_current = 1.0
+
         # If either ws_step or wd_step is None, set it to the current step
         if ws_step is None:
-            if len(self.wind_speeds) >= 2:
-                ws_step = self.wind_speeds[1] - self.wind_speeds[0]
-            else:  # wind rose will have only a single wind speed, and we assume a ws_step of 1
-                ws_step = 1.0
+            ws_step = ws_step_current
         if wd_step is None:
-            if len(self.wind_directions) >= 2:
-                wd_step = self.wind_directions[1] - self.wind_directions[0]
-            else:  # wind rose will have only a single wind direction, and we assume a wd_step of 1
-                wd_step = 1.0
+            wd_step = wd_step_current
 
-        # Set up the new wind direction and wind speed bins
-        new_wind_directions = np.arange(
-            self.wind_directions[0], self.wind_directions[-1] + wd_step / 2.0, wd_step
-        )
-        new_wind_speeds = np.arange(
-            self.wind_speeds[0], self.wind_speeds[-1] + ws_step / 2.0, ws_step
-        )
+        # Identify the covered range of wind directions
+        wd_range_min_current = np.min(self.wind_directions) - wd_step_current / 2.0
+        wd_range_max_current = np.max(self.wind_directions) + wd_step_current / 2.0
+
+        # Look for unlikely case where for example wind directions are 8, 28, ... 358
+        if wd_range_max_current > 360:
+            raise ValueError(
+                "Cannot upsample wind rose for case when wind directions are defined"
+                " such that 0 degrees is included by bins to the left of 0 degrees. "
+            )
+
+        # Identify the new minimum wind direction
+        wd_min_new = wd_range_min_current + wd_step / 2.0
+        wd_max_new = wd_range_max_current - wd_step / 2.0
+
+        new_wind_directions = np.arange(wd_min_new, wd_max_new + wd_step / 2.0, wd_step)
+
+        # Set up the new wind speeds
+        ws_range_min_current = np.min(self.wind_speeds) - ws_step_current / 2.0
+        ws_range_max_current = np.max(self.wind_speeds) + ws_step_current / 2.0
+        ws_min_new = ws_range_min_current + ws_step / 2.0
+        ws_max_new = ws_range_max_current - ws_step / 2.0
+
+        # Force the new ws_min to 0 if negative
+        if ws_min_new < 0:
+            ws_min_new = 0.0
+
+        new_wind_speeds = np.arange(ws_min_new, ws_max_new + ws_step / 2.0, ws_step)
```

Comment:
Could this be strange if `ws_min_new` gets forced to zero in the lines above? All bins would be shifted? Perhaps we need to create `new_wind_speeds` by stepping backwards from `ws_max_new` rather than forwards from `ws_min_new`?

Reply:
I think it's ok, there's not that many ways to set the wind speeds, and less that include 0 so I wrote a few test cases into tests and couldn't see how this gets us into trouble